cni/pkg/nodeagent/informers.go

```diff
@@ -0,0 +1,259 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodeagent
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+
+	"istio.io/istio/cni/pkg/util"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/monitoring"
+)
+
+var (
+	eventTypeTag = monitoring.CreateLabel("type")
+	EventTotals  = monitoring.NewSum(
+		"events_total",
+		"The total number of reconcile events.",
+	)
+)
+
+type K8sHandlers interface {
+	GetPodIfAmbient(podName, podNamespace string) (*corev1.Pod, error)
+	GetAmbientPods() []*corev1.Pod
+	Start()
+}
+
+type InformerHandlers struct {
+	ctx             context.Context
+	dataplane       MeshDataplane
+	systemNamespace string
+
+	queue      controllers.Queue
+	pods       kclient.Client[*corev1.Pod]
+	namespaces kclient.Client[*corev1.Namespace]
+}
+
+func setupHandlers(ctx context.Context, kubeClient kube.Client, dataplane MeshDataplane, systemNamespace string) *InformerHandlers {
+	s := &InformerHandlers{ctx: ctx, dataplane: dataplane, systemNamespace: systemNamespace}
+	s.queue = controllers.NewQueue("ambient",
+		controllers.WithGenericReconciler(s.reconcile),
+		controllers.WithMaxAttempts(5),
+	)
+	// We only need to handle pods on our node
+	s.pods = kclient.NewFiltered[*corev1.Pod](kubeClient, kclient.Filter{FieldSelector: "spec.nodeName=" + NodeName})
+	s.pods.AddEventHandler(controllers.FromEventHandler(func(o controllers.Event) {
+		s.queue.Add(o)
+	}))
+
+	// Namespaces could be anything though, so we watch all of those
+	s.namespaces = kclient.New[*corev1.Namespace](kubeClient)
+	s.namespaces.AddEventHandler(controllers.FromEventHandler(func(o controllers.Event) {
```

Comment:
This model doesn't seem quite right. Was it intentionally changed from the old behavior (which, IMO, is the right way). We are putting it in the queue, then the queue action is to add more items to the queue. The NS queue handler is syncronous so there is no need for a queue anyways

Typically, in the EventHandler we would find what to put on the queue. So `{pod.List(); enqueue each}`

Reply:
This is, AFAICT, more or less how the current informer in `master` is doing it: https://github.com/istio/istio/blob/master/cni/pkg/ambient/informers.go#L46

If we can do it better, happy to rework in a followup.

Reply:
I think the only reason for doing this was to avoid using the `fieldSelector == thisNodeName` filter for the namespace watch, but I would assume namespace updates would ignore node-specific filters, so they can be consolidated.

Reply:
Fixed